fearless/Modules/Staking/StakingRebondSetup/StakingRebondSetupPresenter.swift

```diff
@@ -0,0 +1,217 @@
+            view?.didReceiveFee(viewModel: feeViewModel)
+        } else {
+            view?.didReceiveFee(viewModel: nil)
+        }
+    }
+
+    private func provideAssetViewModel() {
+        let viewModel = balanceViewModelFactory.createAssetBalanceViewModel(
+            inputAmount ?? 0.0,
+            balance: unbonding,
+            priceData: priceData
+        )
+
+        view?.didReceiveAsset(viewModel: viewModel)
+    }
+}
+
+extension StakingRebondSetupPresenter: StakingRebondSetupPresenterProtocol {
+    func selectAmountPercentage(_ percentage: Float) {
+        if let unbonding = unbonding {
+            inputAmount = unbonding * Decimal(Double(percentage))
+            provideInputViewModel()
+            provideAssetViewModel()
+        }
+    }
+
+    func updateAmount(_ amount: Decimal) {
+        inputAmount = amount
+        provideAssetViewModel()
+    }
+
+    func proceed() {
+        let locale = view?.localizationManager?.selectedLocale ?? Locale.current
+        DataValidationRunner(validators: [
+            dataValidatingFactory.canRebond(amount: inputAmount, unbonding: unbonding, locale: locale),
+
+            dataValidatingFactory.has(fee: fee, locale: locale, onError: { [weak self] in
+                self?.interactor.estimateFee()
+            }),
+
+            dataValidatingFactory.canPayFee(balance: balance, fee: fee, locale: locale),
+
+            dataValidatingFactory.has(
+                controller: controller,
+                for: stashItem?.controller ?? "",
+                locale: locale
+            ),
+
+            dataValidatingFactory.electionClosed(electionStatus, locale: locale)
+        ]).runValidation { [weak self] in
+            if let amount = self?.inputAmount {
+                self?.wireframe.proceed(view: self?.view, amount: amount)
+            } else {
+                self?.logger?.warning("Missing amount after validation")
+            }
+        }
+    }
+
+    func close() {
+        wireframe.close(view: view)
+    }
+
+    func setup() {
+        provideInputViewModel()
+        provideFeeViewModel()
+        provideAssetViewModel()
+
+        interactor.setup()
+    }
+}
+
+extension StakingRebondSetupPresenter: StakingRebondSetupInteractorOutputProtocol {
+    func didReceiveFee(result: Result<RuntimeDispatchInfo, Error>) {
+        switch result {
+        case let .success(dispatchInfo):
+            if let fee = BigUInt(dispatchInfo.fee) {
+                self.fee = Decimal.fromSubstrateAmount(fee, precision: chain.addressType.precision)
+            }
+
+            provideFeeViewModel()
+        case let .failure(error):
+            logger?.error("Did receive fee error: \(error)")
+        }
+    }
+
+    func didReceiveStakingLedger(result: Result<DyStakingLedger?, Error>) {
+        switch result {
+        case let .success(stakingLedger):
+            if let stakingLedger = stakingLedger {
+                unbonding = Decimal.fromSubstrateAmount(
```

Comment:
It is bad. What if active era comes before staking info. I suggest just saving the staking ledger and calculate unbonding on the fly.

Reply:
fixed